### ISSUE TITLE
Some commands not recognized for clients connected to a server

### ISSUE THREAD

**Issue by Cervator** [MovingBlocks/Terasology#1619]
In a somewhat odd twist after fixing "teleport" in #1618 it now works in single player, yay! But weirdly the command doesn't even appear to _exist_ when connected as a multiplayer client. Same for a few other commands like health and healthMax, yet certain other commands are fine.

A player hosting a server with the normal game client can also use those commands normally, even with another player connected being told they don't exist. 

Maybe it relates to where commands are defined - maybe some classes aren't processed the same when a client joins?

We probably should check to see if this is the case with other commands too.
##

**Comment by msteiger** [comment]
It seems like you need *cheat permissions* for all of them. Is it just that? Ping @flo

**Comment by Cervator** [comment]
Well, no, then you'd get an error message saying you don't have permissions to run that command. This is like they're not even getting registered in the first place :-)

**Comment by immortius** [comment]
The system providing the command is probably set as server-side only. This
means it isn't created client side and thus the commands are not discovered.

**Comment by HashCode55** [comment]
@Cervator I'm pretty sure the the files which contain this code are not being included on the client side. Because it gives an error 'Unknown command'.... and I traced this back and it can happen when the command hasn't been defined. I spent a day reading the code and I have an intuition that it has to do something with 'Loadprocesses.java', it is the only file where client and the host get differentiated for the first time....

**Comment by flo** [comment]
I opened the call hirachy of the runOnServer method of the Command annotation.

It gets used by method refferingTo in MethodCommaned. It passes the value as argument to
MethodCommand which passes it to the super constructor AbstractCommand. The constructor of AbstractCommand uses it to set a field called runOnServer. The field gets exposed via a isRunOnServer method.  @HashCode55  I suggest you continue your bug hunt at the place where isRunOnServer of AbsttractCommand gets called.

I suggest that you read https://github.com/MovingBlocks/Terasology/wiki/Events-and-systems for background knowledge about the events and systems. Especially about what it says about RegisterMode.

**Comment by HashCode55** [comment]
@flo 
I traced it back from where you said but it seems to be working fine. isRunOnServer() is called in consoleimpl().java. If it it true and the application has authority (the server) it will execute the command right away, but if not (the remote client in this case), it will send at event and another java file named Commandsystem.java will receive the event and apply a filter for authority and execute the command. 

But the problem is not here I guess, it says 'unknown command', which means that the commands are not there in 'commandregistry', which means the classes are missing.

**Comment by flo** [comment]
I searched for the logged error message and found it in ConsoleImpl#execute.It gets printed when getCommand returns null. commandRegistry does a lookup in the map field commandRegistry.

I followed that trail till I found out that MethodCommand.registerAvailable is repsonsible for registering commands of a system. It gets called when a system gets initialized

loadSystems of ComponentSystemManager creates systems only when they have been declared to run in that environement.

In other words: Commands should never be declared in a system with RegisterMode AUTHORITY as otherwise they will just work when ran as authority. Since we have a permission system now, any command in a class with RegisterMode AUTHORITY is a bug. 

I suggest we let registerAvailable fail with an exception whenever a command gets found in a system with RegisterMode AUTHORITY. That would prevent such bugs in future as the one who implements it wrongly will notice it instantly what he has to change.

Offending commands should be moved into a separate client system with the correct RegisterMode. Client systems get usually called *ClientSystem. The teleport command could be moved to MovementDebugCommands..

**Comment by HashCode55** [comment]
Wow! That totally makes sense and connects the broken threads.

**Comment by HashCode55** [comment]
@Flo 
Do you think there will be any logical problem if we shift all commands from `healthauthoritysystem` to `healthclientsystem`?

**Comment by Josharias** [comment]
Heyo @HashCode55,  Sorry I am late to the game on this one.  I would expect the health system to not function in a multiplayer scenario if you move the code to the health client system.

I would suggest creating a common system that has the ```runOnServer = true``` commands in them.  Possibly ```HealthCommonSystem```?  And then register this system as ```RegisterMode.ALWAYS```.

Flo's suggestion about logging errors when a ```runOnServer = true```  command is registered on an ```RegisterMode.AUTHORITY``` is an excellent one.  Thanks for running this problem down.

**Comment by flo** [comment]
@Josharias  hmm yeah you are right, the command need currently to be defined in system with register mode ALWAYS. Otherwise the server won't be able to process the commands...

Since client systems don't execute the commands runOnSErver=true commands, it would propably make sense to add some logic to make it possible to clients to find command in server systems that have runOnServer=true.

Then we could let all commands on authority systems be automatically runOnServer commands.

Option 1.)
Instead of registering the commands for registered systems, commands could be registered by looping all known systems:
```
        for (Class<?> type : environment.getTypesAnnotatedWith(RegisterSystem.class)) {
```
this solution does however not work for server only modules.

So I would suggest option 2.)
The client always sends all commands to the server, who interprets them. The command completion gets also done one the server. The client sends the unfinished command to the server and the server replies with a suggestion list.

of course there is also option 3.)
We could require all commands with runOnServer=true to be defined in systems with register mode ALWAYS and throw an error otherwise. 

I would prefer option 2, as I think a clear seperation between server and client helps.

**Comment by flo** [comment]
@HashCode55 I updated the previous comment

**Comment by HashCode55** [comment]
@flo 
The commands are actually getting recognized on both client and server if ran on register mode CLIENT.

 But yeah currently client doesn't execute the runOnServer commands and sends it to the server for execution. And that is why there is a lag in between the command and the result, of like 8-9 seconds.
Option 2 will also prevent this unwanted lag....

**Comment by flo** [comment]
hmm, I tried moving a runOnServer command to the client and it did not seem to work. (It would also be logic not to work when the system that is supposed to run the command has not been created on the server).

since you are currently online, lets meet in the IRC channel?

**Comment by flo** [comment]
There seem to be some inconsistency in argument checking, that got me confused>

Commands (at least those with runOnServer=true) that get run on on authority get executed even if additional arguments get added to the command string. When to few arguments are provided an error gets logged.

If remote client commands however don't have the correct argument count, the server ignores them without any kind of user feedback.

**Comment by flo** [comment]
As I commented in #2217:

The health modifying commands need to run on the server. It is strange that we let commands define in a client system that with runOnServer=true as they need then to call functionality that is in the authority system.

We should probably change it and make it so that commands that get executed on the server can be (only) defined in systems with authority(or the always mode).

That means that the client won't know commands that are supposed to run on the server. So we would have to change the client handling to forwards unknown commands to the server in the hope that they can be executed there.

**Comment by MaloJaffre** [comment]
We could also transmit a list of server commands to the client for reducing error latency and server load.

**Comment by flo** [comment]
The command completion does not only allow you to complete the command name but also arguments. So the completion would be something where the client has to ask the server too.

**Comment by HashCode55** [comment]
@flo but won't that increase the load on the server because any of the commands which are unknown to the client, irrespective of whether its run on server or not, will be sent to the server to check if it has been defined there or not.

**Comment by flo** [comment]
The server sends millions of blocks to the server. A single command wouldn't make a big difference. The only problematic thing is delay during command completion.

Maybe we should let the client simply search all known systems for commands. Then we just have the limitation that commands in sever only modules can't be found.

**Comment by flo** [comment]
`loadSystems` in `ComponentSystemManager` determines first a list of all known systems and then determines which need to be loaded.

**Comment by flo** [comment]
There probably needs to be 2 kinds of command representing objects to be stored:
1. Command meta data for all known commands that get used for command completion on the client.
2. Executable command objects that get gathered only for  the systems that are active. e.g. for a object that gets only executed on the server, only the server would have the executable command object.

**Comment by HashCode55** [comment]
Yeah but this thing is already there, the commands which are known on the client side get registered in the `commandRegistry`. So this means the client side already searches the commands. But the commands which get misspelled by me will be sent to server to check....

**Comment by flo** [comment]
not sure if your reply was for my latest post, but: Currently we have just 1 command representing object that exists only for the active systems. The client does not know about command that are defined in authority only systems.

**Comment by HashCode55** [comment]
Oh sorry, It was for the previous post :P . 
Yeah I think this is a nice idea... (of having the metadata stored)

**Comment by portokaliu** [comment]
I've tried separating the commands to a separate HealthCommands class and use that one in a way aking to BlockCommands and everything worked out just fine. (on client as well as host)

I saw the issue is still present, so I'd like to know if you guys think that's a good way to do it or should we find another way? I for one consider it pretty straightforward.

**Comment by Cervator** [comment]
Closing as fixed by #2600 :-)
